### PR TITLE
[Android] Respect "filename*" parameter in the field Content-Disposition when detecting filenames for downloading.

### PR DESCRIPTION
**Pull request by UNIDY2002**
According to [Section 4.3 of RFC6266](https://www.rfc-editor.org/rfc/rfc6266#section-4.3), filenames with non-ASCII characters may be specified with a `"filename*"` parameter.

This library uses `android.webkit.URLUtil.guessFileName` to detect filenames. However, the API is a simplistic implementation of the RFC, and thus cannot understand the `"filename*"` parameter, which leads to issue #2353.

I modified the implementation of `android.webkit.URLUtil` and added a naive support for the `"filename*"` parameter. Besides, I added a test case in `example/examples/Downloads.tsx` to cover this case.

NOTE that my modified implementation is still too simple, and may fail to cover [every corner case](http://test.greenbytes.de/tech/tc2231/) of the RFC. Nevertheless, since filename detection is not the focus of this library, the modification shall be valuable as long as it understands the `"filename*"` parameter in most cases.

Looking forward to your revision.

### CODE REVIEW

File: android/src/main/java/com/reactnativecommunity/webview/URLUtil.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (C) 2006 The Android Open Source Project
```
Comment (TheAlmightyBob) [inline android/src/main/java/com/reactnativecommunity/webview/URLUtil.java:2]: Please clarify what exactly is modified in this file.

File: example/examples/Downloads.tsx [hunk 2]
```diff
@@ -21,6 +21,8 @@ const HTML = `
   </head>
   <body>
     <a href="https://www.7-zip.org/a/7za920.zip">Example zip file download</a>
+    <br>
+    <a href="http://test.greenbytes.de/tech/tc2231/attwithisofn2231iso.asis">Download file with non-ascii filename</a>
```
Comment (TheAlmightyBob) [inline example/examples/Downloads.tsx:25]: Would be helpful to say in this text (or at the very least a code comment) what the expected filename is

**Review by TheAlmightyBob — CHANGES_REQUESTED** [review]

File: android/src/main/java/com/reactnativecommunity/webview/URLUtil.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            Matcher m = CONTENT_DISPOSITION_PATTERN.matcher(contentDisposition);
+            if (m.find()) {
+                if (m.group(3) != null && m.group(4) != null) {
+                    try {
+                        return URLDecoder.decode(m.group(4), m.group(3));
+                    } catch (UnsupportedEncodingException e) {
+                        // Skip the ext-parameter as the encoding is unsupported
+                    }
+                }
+                return m.group(2);
```
Comment (TheAlmightyBob) [inline android/src/main/java/com/reactnativecommunity/webview/URLUtil.java:161]: I like the summary change list comment you put at the top, but could you also add something briefly here to say what this is doing? (e.g., what's the significance of these groups, how does this relate to "filename*", why is decoding required..)

**Review by TheAlmightyBob — CHANGES_REQUESTED** [review]
Much better, thanks, just one more thing..

File: android/src/main/java/com/reactnativecommunity/webview/URLUtil.java (hunk 3 above)
Comment (UNIDY2002) [reply android/src/main/java/com/reactnativecommunity/webview/URLUtil.java]: Thanks for your feedback.

According to the RFC, `m.group(4)` refers to the percent-encoded filename, and the charset is specified in `m.group(3)`; `m.group(2)` is the fallback filename.

I have also left some comments in the source code.

File: android/src/main/java/com/reactnativecommunity/webview/URLUtil.java [hunk 4]
```diff
@@ -135,25 +135,32 @@ public static final String guessFileName(
...
-     * Unfortunately some servers do not quote the value so to maintain
-     * consistent behaviour with other browsers, we allow unquoted values too.
      */
     static String parseContentDisposition(String contentDisposition) {
         try {
+            // The regex attempts to match the following pattern:
+            //     attachment; filename="(Group 2)"; filename*=(Group 3)'(lang)'(Group 4)
+            // (Group 1 refers to the quotation marks around Group 2)
+            //
+            // Test cases can be found at http://test.greenbytes.de/tech/tc2231/
```
Comment (TheAlmightyBob) [inline android/src/main/java/com/reactnativecommunity/webview/URLUtil.java:156]: Thanks for including that! Seems to have other cases that are not supported, such as `filename=foo.html` (no quotes) or breaking things up with `filename*1`/`filename*2`...

But it doesn't look like your changes break support for a previously-working case, so that should be good.

**Review by TheAlmightyBob — APPROVED** [review]

File: android/src/main/java/com/reactnativecommunity/webview/URLUtil.java (hunk 4 above)
Comment (UNIDY2002) [reply android/src/main/java/com/reactnativecommunity/webview/URLUtil.java:156]: Yes, that's the case. If there remain issues in the future regarding filename guessing, I would like to switch to a more robust implementation.

Thank you!
